File: 1675-minimize-deviation-in-array/1675-minimize-deviation-in-array.cpp
class Solution {
public:
    int minimumDeviation(vector<int>& nums) {
        set<int> s;
        for (int &num : nums) {
            s.insert(num % 2 ? num*2 : num);
        }
        int res = *s.rbegin() - *s.begin();
        while (*s.rbegin() % 2 == 0) {
            s.insert(*s.rbegin() / 2);
            s.erase(*s.rbegin());
            res = min(res, *s.rbegin() - *s.begin());
        }
        return res;
    }
};

File: 1675-minimize-deviation-in-array/1675-minimize-deviation-in-array.java
class Solution {
    public int minimumDeviation(int[] nums) {
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        int n = nums.length, mi = Integer.MAX_VALUE, res = Integer.MAX_VALUE;
        for (int num : nums) {
            if (num % 2 == 1)
                num *= 2;
            pq.add(-num);
            mi = Math.min(mi,num);
        }
        
        while (true) {
            int a = -pq.poll();
            res = Math.min(res, a - mi);
            if (a % 2 == 1)
                break;
            mi = Math.min(mi, a / 2);
            pq.add(-a/ 2);
        }
        return res;
    }
}